src/ragas/optimizers/genetic.py

```diff
@@ -0,0 +1,733 @@
+
+        if self.llm is None:
+            raise ValueError("No llm provided for optimization.")
+
+        parents = ParentPrompts(parent_1=parent_1, parent_2=parent_2)
+        offspring = await self.cross_over_prompt.generate(
+            data=parents, llm=self.llm, callbacks=callbacks
+        )
+        return offspring.instruction
+
+    def _set_instructions(self, candidates: t.Dict[str, str]):
+        if self.metric is None:
+            raise ValueError("No metric provided for optimization.")
+        prompts = self.metric.get_prompts()
+        for key, val in candidates.items():
+            prompts[key].instruction = val
+        self.metric.set_prompts(**prompts)
+
+    def feedback_mutation(
+        self,
+        candidates: t.List[t.Dict[str, str]],
+        dataset: SingleMetricAnnotation,
+        sample_size: int,
+        run_config: t.Optional[RunConfig] = None,
+        batch_size: t.Optional[int] = None,
+        callbacks: t.Optional[Callbacks] = None,
+        raise_exceptions: bool = True,
+        pbar: t.Optional[tqdm] = None,
+    ) -> t.List[t.Dict[str, str]]:
+
+        if self.metric is None:
+            raise ValueError("No metric provided for optimization.")
+
+        feedback_rm, feedback_grp = new_group(
+            name="Feedback mutation",
+            inputs={"candidates": candidates},
+            callbacks=callbacks,
+        )
+        improved_candidates = []
+        dataset = dataset.filter(lambda x: x["is_accepted"])
+        for candidate in candidates:
+            candidate_rm, candidate_grp = new_group(
+                name="Candidate feedback mutation",
+                inputs={"candidate": candidate},
+                callbacks=feedback_grp,
+            )
+            dataset_sample = dataset.sample(sample_size, stratify_key="metric_output")
+            batch, target = self._get_evaluation_dataset(dataset_sample)
+            results = self.evaluate_candidate(
+                candidate=candidate,
+                eval_dataset=batch,
+                run_config=run_config,
+                batch_size=batch_size,
+                callbacks=candidate_grp,
+                raise_exceptions=raise_exceptions,
+                run_id=candidate_rm.run_id,
+                pbar=pbar,
+            )
+
+            exec = Executor(
+                desc="Getting feedbacks",
+                raise_exceptions=raise_exceptions,
+                run_config=run_config,
+                keep_progress_bar=False,
+                batch_size=batch_size,
+                pbar=pbar,
+            )
+            exec.submit(
+                self._feedback_mutation,
+                candidate=candidate,
+                dataset=dataset_sample,
+                results=results,
+                target=target,
+                callbacks=candidate_grp,
+            )
+            try:
+                improved_candidate = exec.results()
+                improved_candidates.append(improved_candidate[0])
```

Comment:
shouldn't we submit all the jobs and then wait for the results outside the loop?